frontend/src/metabase/dashboard/actions/data-fetching-typed.ts

```diff
@@ -203,10 +239,341 @@ export const markCardAsSlow = createAction(MARK_CARD_AS_SLOW, (card: Card) => {
 });
 
 export const SET_DOCUMENT_TITLE = "metabase/dashboard/SET_DOCUMENT_TITLE";
+
 export const setDocumentTitle = createAction<string>(SET_DOCUMENT_TITLE);
 
 export const SET_SHOW_LOADING_COMPLETE_FAVICON =
   "metabase/dashboard/SET_SHOW_LOADING_COMPLETE_FAVICON";
 export const setShowLoadingCompleteFavicon = createAction<boolean>(
   SET_SHOW_LOADING_COMPLETE_FAVICON,
 );
+
+export const FETCH_DASHBOARD_CARD_DATA =
+  "metabase/dashboard/FETCH_DASHBOARD_CARD_DATA";
+
+export const fetchDashboardCardDataAction = createAction<{
+  currentTime: number;
+  loadingIds: DashCardId[];
+}>(FETCH_DASHBOARD_CARD_DATA);
+
+export const FETCH_CARD_DATA_PENDING =
+  "metabase/dashboard/FETCH_CARD_DATA/pending";
+export const fetchCardDataPending = createAction(
+  FETCH_CARD_DATA_PENDING,
+  (dashcard_id: DashCardId) => ({ payload: { dashcard_id } }),
+);
+
+export const CANCEL_FETCH_CARD_DATA =
+  "metabase/dashboard/CANCEL_FETCH_CARD_DATA";
+
+const cardDataCancelDeferreds: Record<
+  `${DashCardId},${DashboardCard["card_id"]}`,
```

Comment:
I like how this describes how the key shape looks like in the type signature itself, nice!